File: ChatBot/ChatBot/Conversation.m
#import "Conversation.h"

@implementation Conversation

+ (instancetype)conversationWithTitle:(NSString *)title {
    Conversation *conversation = [[Conversation alloc] init];
    conversation.title = title;
    conversation.createdAt = [NSDate date];
    conversation.messages = [NSMutableArray array];
    return conversation;
}

- (void)addMessage:(Message *)message {
    [self.messages addObject:message];
    [self saveToUserDefaults];
}

- (void)saveToUserDefaults {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSMutableArray *conversations = [[defaults arrayForKey:@"conversations"] mutableCopy] ?: [NSMutableArray array];
    
    // 将当前对话转换为可存储的格式
    NSMutableDictionary *conversationDict = [NSMutableDictionary dictionary];
    conversationDict[@"title"] = self.title;
    conversationDict[@"createdAt"] = self.createdAt;
    
    NSMutableArray *messagesArray = [NSMutableArray array];
    for (Message *message in self.messages) {
        [messagesArray addObject:@{
            @"content": message.content,
            @"isUser": @(message.isUser),
            @"timestamp": message.timestamp
        }];
    }
    conversationDict[@"messages"] = messagesArray;
    
    // 查找并更新或添加对话
    BOOL found = NO;
    for (NSInteger i = 0; i < conversations.count; i++) {
        NSDictionary *dict = conversations[i];
        if ([dict[@"title"] isEqualToString:self.title]) {
            conversations[i] = conversationDict;
            found = YES;
            break;
        }
    }
    
    if (!found) {
        [conversations addObject:conversationDict];
    }
    
    [defaults setObject:conversations forKey:@"conversations"];
    [defaults synchronize];
}

+ (NSArray<Conversation *> *)loadAllConversations {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSArray *conversationsArray = [defaults arrayForKey:@"conversations"] ?: @[];
    
    NSMutableArray *conversations = [NSMutableArray array];
    for (NSDictionary *dict in conversationsArray) {
        Conversation *conversation = [[Conversation alloc] init];
        conversation.title = dict[@"title"];
        conversation.createdAt = dict[@"createdAt"];
        
        NSMutableArray *messages = [NSMutableArray array];
        for (NSDictionary *messageDict in dict[@"messages"]) {
            Message *message = [Message messageWithContent:messageDict[@"content"]
                                                  isUser:[messageDict[@"isUser"] boolValue]];
            message.timestamp = messageDict[@"timestamp"];
            [messages addObject:message];
        }
        conversation.messages = messages;
        
        [conversations addObject:conversation];
    }
    
    // 按创建时间排序
    return [conversations sortedArrayUsingComparator:^NSComparisonResult(Conversation *conv1, Conversation *conv2) {
        return [conv2.createdAt compare:conv1.createdAt];
    }];
}

+ (void)deleteConversation:(Conversation *)conversation {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSMutableArray *conversations = [[defaults arrayForKey:@"conversations"] mutableCopy] ?: [NSMutableArray array];
    
    [conversations removeObjectAtIndex:[conversations indexOfObjectPassingTest:^BOOL(NSDictionary *dict, NSUInteger idx, BOOL *stop) {
        return [dict[@"title"] isEqualToString:conversation.title];
    }]];
    
    [defaults setObject:conversations forKey:@"conversations"];
    [defaults synchronize];
}

@end 

File: ChatBot/ChatBot/Conversation.swift
import Foundation

class Conversation {
    let title: String
    var createdAt: Date
    var messages: [Message]
    
    init(title: String) {
        self.title = title
        self.createdAt = Date()
        self.messages = []
    }
    
    func addMessage(_ message: Message) {
        messages.append(message)
        saveToUserDefaults()
    }
    
    func saveToUserDefaults() {
        let defaults = UserDefaults.standard
        var conversations = defaults.array(forKey: "conversations") as? [[String: Any]] ?? []
        
        // 将当前对话转换为可存储的格式
        var conversationDict: [String: Any] = [:]
        conversationDict["title"] = title
        conversationDict["createdAt"] = createdAt
        
        let messagesArray = messages.map { message -> [String: Any] in
            return [
                "content": message.content,
                "isUser": message.isUser,
                "timestamp": message.timestamp
            ]
        }
        conversationDict["messages"] = messagesArray
        
        // 查找并更新或添加对话
        if let index = conversations.firstIndex(where: { ($0["title"] as? String) == title }) {
            conversations[index] = conversationDict
        } else {
            conversations.append(conversationDict)
        }
        
        defaults.set(conversations, forKey: "conversations")
        defaults.synchronize()
    }
    
    static func loadAllConversations() -> [Conversation] {
        let defaults = UserDefaults.standard
        let conversationsArray = defaults.array(forKey: "conversations") as? [[String: Any]] ?? []
        
        let conversations = conversationsArray.compactMap { dict -> Conversation? in
            guard let title = dict["title"] as? String,
                  let createdAt = dict["createdAt"] as? Date,
                  let messagesArray = dict["messages"] as? [[String: Any]] else {
                return nil
            }
            
            let conversation = Conversation(title: title)
            conversation.createdAt = createdAt
            
            conversation.messages = messagesArray.compactMap { messageDict -> Message? in
                guard let content = messageDict["content"] as? String,
                      let isUser = messageDict["isUser"] as? Bool,
                      let timestamp = messageDict["timestamp"] as? Date else {
                    return nil
                }
                
                let message = Message(content: content, isUser: isUser)
                message.timestamp = timestamp
                return message
            }
            
            return conversation
        }
        
        // 按创建时间排序
        return conversations.sorted { $0.createdAt > $1.createdAt }
    }
    
    static func deleteConversation(_ conversation: Conversation) {
        let defaults = UserDefaults.standard
        var conversations = defaults.array(forKey: "conversations") as? [[String: Any]] ?? []
        
        conversations.removeAll { ($0["title"] as? String) == conversation.title }
        
        defaults.set(conversations, forKey: "conversations")
        defaults.synchronize()
    }
} 